skypeweb/skypeweb_login.c

```diff
@@ -317,3 +317,250 @@ skypeweb_refresh_token_login(SkypeWebAccount *sa)
 	
 	purple_connection_update_progress(sa->pc, _("Authenticating"), 2, 4);
 }
+
+
+static void
+skypeweb_login_did_got_api_skypetoken(PurpleHttpConnection *http_conn, PurpleHttpResponse *response, gpointer user_data)
+{
+	SkypeWebAccount *sa = user_data;
+	const gchar *data;
+	gsize len;
+	JsonParser *parser = NULL;
+	JsonNode *node;
+	JsonObject *obj;
+	gchar *error = NULL;
+	PurpleConnectionError error_type = PURPLE_CONNECTION_ERROR_NETWORK_ERROR;
+
+	data = purple_http_response_get_data(response, &len);
+
+	parser = json_parser_new();
+	if (!json_parser_load_from_data(parser, data, len, NULL)) {
+		goto fail;
+	}
+
+	node = json_parser_get_root(parser);
+	if (node == NULL || json_node_get_node_type(node) != JSON_NODE_OBJECT) {
+		goto fail;
+	}
+	obj = json_node_get_object(node);
+
+	if (!json_object_has_member(obj, "skypetoken")) {
+		JsonObject *status = json_object_get_object_member(obj, "status");
+		if (status) {
+			//{"status":{"code":40120,"text":"Authentication failed. Bad username or password."}}
+			error = g_strdup_printf(_("Login error: %s (code %d)"),
+				json_object_get_string_member(status, "text"),
+				json_object_get_int_member(status, "code")
+			);
+			error_type = PURPLE_CONNECTION_ERROR_AUTHENTICATION_FAILED;
+		}
+		goto fail;

```

Comment:
No, it is not when it's used for error handling in C. It's cleaner than the other alternatives.

I didn't think eion was serious, I'm sure he knows this already. It's used all over the place in libpurple, the linux kernel, etc.

https://stackoverflow.com/questions/788903/valid-use-of-goto-for-error-management-in-c

https://softwareengineering.stackexchange.com/questions/154974/is-this-a-decent-use-case-for-goto-in-c

http://eli.thegreenplace.net/2009/04/27/using-goto-for-error-handling-in-c

https://blog.regehr.org/archives/894

http://web.archive.org/web/20130411141953/http://kerneltrap.org/node/553





http://eli.thegreenplace.net/2009/04/27/using-goto-for-error-handling-in-c